packages/yarnpkg-cli/sources/lib.ts

```diff
@@ -89,12 +89,17 @@ function runYarnPath(cli: YarnCli, argv: Array<string>, {yarnPath}: {yarnPath: P
   return 0;
 }
 
-function checkCwd(argv: Array<string>): [PortablePath, Array<string>] {
-  if (argv.length >= 2 && argv[0] === `--cwd`)
-    return [npath.toPortablePath(npath.resolve(argv[1])), argv.slice(2)];
-
-  if (argv.length >= 1 && argv[0].startsWith(`--cwd=`))
-    return [npath.toPortablePath(npath.resolve(argv[0].slice(6))), argv.slice(1)];
+function checkCwd(cli: YarnCli, argv: Array<string>) {
+  let cwd: PortablePath | null = null;
+
+  let postCwdArgv = argv;
+  if (argv.length >= 2 && argv[0] === `--cwd`) {
+    cwd = npath.toPortablePath(argv[1]);
+    postCwdArgv = argv.slice(2);
+  } else if (argv.length >= 1 && argv[0].startsWith(`--cwd=`)) {
+    cwd = npath.toPortablePath(argv[0].slice(6));
+    postCwdArgv = argv.slice(1);
+  }
 
   cli.defaultContext.cwd = cwd ?? ppath.cwd();
```

Comment:
You removed the `resolve` calls, which are important because everything breaks if `this.context.cwd` isn't an absolute path.
```suggestion
  cli.defaultContext.cwd = ppath.resolve(cwd) ?? ppath.cwd();
```